pkgs/fc/sensuplugins/fc/sensuplugins/http.py

```diff
@@ -0,0 +1,131 @@
+import requests
+from requests_toolbelt.adapters.source import SourceAddressAdapter
+
+
+def err(msg):
+    print(msg, file=sys.stderr)
+    sys.exit(1)
+
+
+def unknown(msg):
+    print(f"HTTP UNKNOWN - {msg}")
+    sys.exit(1)
+
+
+def warning(msg):
+    print(f"HTTP WARNING - {msg}")
+    sys.exit(1)
+
+
+def critical(msg):
+    print(f"HTTP CRITICAL - {msg}")
+    sys.exit(1)
+
+
+def ok(msg):
+    print(f"HTTP OK - {msg}")
+
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "-4", "--ipv4", action="store_true", help="Only use IPv4"
+    )
+    parser.add_argument(
+        "-6", "--ipv6", action="store_true", help="Only use IPv6"
+    )
+    parser.add_argument(
+        "-e",
+        "--expect",
+        type=int,
+        action="append",
+        help="Expected status codes",
+    )
+    parser.add_argument(
+        "-f", "--follow", action="store_true", help="Follow HTTP redirects"
+    )
+    parser.add_argument("URL", help="URL to be fetched")
+
+    args = parser.parse_args()
+
+    # forcing a specific address family by binding to its wildcard
+    # address works on linux, but does not seem to work on darwin.
+    source_address = None
+    if args.ipv4 and args.ipv6:
+        err("conflicting arguments, -4 and -6 cannot be specified together")
+    elif args.ipv4:
+        source_address = "0.0.0.0"
+    elif args.ipv6:
+        source_address = "::"
+
+    if args.expect is not None and any(
+        map(lambda n: n < 100 or n > 599, args.expect)
+    ):
+        err("invalid expected status code")
+
+    try:
+        url = urllib.parse.urlsplit(args.URL)
+    except ValueError as ex:
+        err(f"could not parse URL: {ex}")
+
+    if url.scheme != "http" and url.scheme != "https":
+        err(f"unsupported url scheme: {url.scheme}")
+
+    if url.hostname is None:
+        err("empty hostname in URL")
+
+    session = requests.Session()
+    # match monitoring-plugins defaults
+    session.max_redirects = 10
+
+    if url.username is not None:
+        session.auth = (url.username, url.password)
+
+    target = urllib.parse.urlunsplit(
+        (url.scheme, url.hostname, url.path, url.query, "")
+    )
+
+    if source_address is not None:
+        session.mount("http://", SourceAddressAdapter(source_address))
+        session.mount("https://", SourceAddressAdapter(source_address))
+
+    try:
+        resp = session.get(target, allow_redirects=args.follow)
+    except requests.ConnectionError as ex:
+        critical("could not connect to remote host")
+    except requests.TooManyRedirects as ex:
+        warning("maximum redirection depth exceeded")
+    except requests.Timeout as ex:
+        critical("request to remote host timed out")
+
+    if args.expect is None:
+        if resp.status_code >= 500:
+            critical(
+                f"request returned server error status code: {resp.status_code}"
+            )
+        elif resp.status_code >= 400:
+            warning(
```

Comment:
this should be critical, too

Reply:
This is likewise behaviour adopted from monitoring-plugins (https://github.com/monitoring-plugins/monitoring-plugins/blob/v2.3.5/plugins/check_http.c#L1217-L1221). Do we want to keep this behaviour too?

Reply:
In general a 400 error code indicates an error on the client side. However: in monitoring we create specific requests that are supposed NOT to be in error and thus a change on the server side (e.g. a URL going away) can cause an error that otherwise would correctly be an error on the client side. In this case it becomes: either your monitoring is out of date, or we accidentally broke something on the server side. We can't distinguish this and should err on the side of caution and consider this critical.

Reply:
Fixed in [48b0f76](https://github.com/flyingcircusio/fc-nixos/pull/940/commits/48b0f764d57285d3463383612333e3a647ce217a).